operator/internal/lifecycle/client.go

```diff
@@ -0,0 +1,387 @@
+	}
+	list, ok := olist.(client.ObjectList)
+	if !ok {
+		return nil, fmt.Errorf("invalid object list type: %T", object)
+	}
+
+	if err := r.client.List(ctx, list, opts...); err != nil {
+		return nil, fmt.Errorf("listing resources: %w", err)
+	}
+
+	converted := []client.Object{}
+	items := reflect.ValueOf(list).Elem().FieldByName("Items")
+	for i := 0; i < items.Len(); i++ {
+		item := items.Index(i).Addr().Interface().(client.Object)
+		item.GetObjectKind().SetGroupVersionKind(*kind)
+		converted = append(converted, item)
+	}
+
+	return sortCreation(converted), nil
+}
+
+// listAllOwnedResources lists all resources owned by a given cluster, optionally including node pools.
+func (r *ResourceClient[T, U]) listAllOwnedResources(ctx context.Context, owner U, includeNodePools bool) ([]client.Object, error) {
+	resources := []client.Object{}
+	for _, resourceType := range append(r.simpleResourceRenderer.WatchedResourceTypes(), &appsv1.StatefulSet{}) {
+		matching, err := r.listResources(ctx, resourceType, client.MatchingLabels(r.ownershipResolver.GetOwnerLabels(owner)))
+		if err != nil {
+			return nil, err
+		}
+		filtered := []client.Object{}
+		for i := range matching {
+			// special case the node pools
+			if includeNodePools || !r.nodePoolRenderer.IsNodePool(matching[i]) {
+				filtered = append(filtered, matching[i])
+			}
+		}
+		resources = append(resources, filtered...)
+	}
+	return resources, nil
+}
+
+// patchOwnedResource applies a patch to a resource owned by the cluster.
+func (r *ResourceClient[T, U]) patchOwnedResource(ctx context.Context, owner U, object client.Object, extraLabels ...map[string]string) error {
+	if err := r.normalize(object, owner, extraLabels...); err != nil {
+		return err
+	}
+	return r.client.Patch(ctx, object, client.Apply, defaultFieldOwner, client.ForceOwnership)
+}
+
+// normalize normalizes an object by setting its labels and owner references. Any labels passed in as `extraLabels`
+// will potentially override those set by the ownership resolver.
+func (r *ResourceClient[T, U]) normalize(object client.Object, owner U, extraLabels ...map[string]string) error {
+	kind, err := getGroupVersionKind(r.scheme, object)
+	if err != nil {
+		return err
+	}
+	mapping, err := getResourceScope(r.mapper, r.scheme, object)
+	if err != nil {
+		return err
+	}
+
+	// This needs to be set explicitly in order for SSA to function properly.
+	// If an initialized pointer to a concrete CR has not specified its GVK
+	// explicitly, SSA will fail.
+	object.GetObjectKind().SetGroupVersionKind(*kind)
+
+	labels := object.GetLabels()
+	if labels == nil {
+		labels = map[string]string{}
+	}
+
+	for name, value := range r.ownershipResolver.AddLabels(owner) {
+		labels[name] = value
+	}
+
+	for _, extra := range extraLabels {
+		for name, value := range extra {
+			labels[name] = value
+		}
+	}
+
+	object.SetLabels(labels)
+
+	if mapping.Name() == meta.RESTScopeNamespace.Name() {
```

Comment:
Should this method be responsible for setting `.Namespace` as well? I'm always a bit back and forth on whether or not it default Namespace is the right choice but I've been hurt by forgetting .Namespace many more times than it getting set to the wrong value.